### PR TITLE
Save all fields to `localStorage`

### LINKED ISSUE #4: Save all fields to `localStorage`
Right now, all inputs are lost if you refresh the page or return later. I think a reasonable implementation would be one such that:

1. Pressing the "Submit" button stores all form inputs in browser localStorage (or perhaps storage is updated on keypress within the input fields)
2. All visible inputs are restored from localStorage on page load
3. There is "Clear" button which resets all fields and deletes the localStorage values (e.g., clear all values and refresh the page)

### PR DESCRIPTION
**Pull request by cristianmacedo**
Closes #4

As highlighted by @chriskthomas in issue #4, right now, all input values are lost upon page refresh or revisiting the page later. To address this, I've implemented a localStorage saving mechanism with the following steps:

1. **LocalStorage Data Storage**
   Upon pressing the "Submit" button, all form values are now saved to a `form` `localStorage` item. This is achieved using the [FormData](https://developer.mozilla.org/en-US/docs/Web/API/FormData) interface, which conveniently captures the key/value pairs of form fields and their corresponding values.

   > To adhere to the [Input HTML Specification](https://html.spec.whatwg.org/multipage/input.html#common-input-element-apis) saying that a file input _"Throws an "InvalidStateError" DOMException if it is set to any value other than the empty string"_ I've excluded file input values from being saved to `localStorage`. A `localStorageIgnore` list has been introduced to handle this exclusion. For the best approach, we could explore dynamically excluding any input field that isn't an instance of a `File` object before saving.

2. **Restoring Saved Values**
   When the page loads, the values stored in the `form` `localStorage` item are automatically populated into their respective input fields based on the `name` property. Additionally, for input fields with the prefix `links` that are not already present in the document, and have an index higher than the current index of existing links, a function is triggered to create the missing input fields. This ensures that additional links saved to `localStorage` are correctly reconstructed upon page load.

3. **Clear Functionality**
   I've implemented a "Clear" button located next to the "Submit" button. This button, when clicked, utilizes the `form.reset()` method to clear all form values. Additionally, any data stored in the `form` `localStorage` item is removed, providing a clean slate for the user.

   > The choice to use `form.reset()` over reloading the entire page was made to maintain a smooth user experience without unnecessary page refreshes.

I welcome feedback and suggestions on the dynamic approach for handling file inputs and any other parts of the implementation. By the way, I've gone ahead and made a few adjustments to variable names and aspects of code-splitting as well as some [JSDoc](https://jsdoc.app) comments to some functions. I hope that's alright – just aiming to enhance the overall structure.

### CODE REVIEW

File: src/api.php [hunk 1]
```diff
@@ -31,49 +31,55 @@
...
-    <img id="userPhoto" src="<?=$user_photo?>" alt="User Photo">
-    <?php }?>
+    <?php if (!empty($user_photo)) { ?>
+        <img id="userPhoto" src="<?= $user_photo ?>" alt="User Photo">
+    <?php } ?>
 
-    <a href="<?=(!empty($_POST["url"]) ? $_POST["url"] : ".")?>">
-        <h1 id="userName"><?=$_POST["name"]?></h1>
+    <a href="<?= (!empty($_POST["url"]) ? $_POST["url"] : ".") ?>">
+        <h1 id="userName"><?= $_POST["name"] ?></h1>
```
[the remark this answers is not in the document]
Comment (cristianmacedo) [reply src/api.php:55]: I think the bot doesn't identify that this heading tag text is filled by PHP code. We already validate that the `name` property is valid and present at form level before submit.

**Comment by mwt** [comment]
Let's give it a try!

**Comment by mwt** [comment]
It's live at the test domain: https://linkfree.fly.dev

It seems to be working. I haven't read the code yet. It seems good.

The preview checkbox doesn't get saved. I think this is good. Is this also explicitly excluded?

**Comment by cristianmacedo** [comment]
HI @mwt, thanks for merging the PR, it seems to be working great on the production environment.

Regarding the `preview` button, from my understanding that's actually a bug. The localStorage script tries to set the `input type="checkbox"` value to "on", but the property that should be updated so that the checkbox turns into, well... checked, is the `checked` property, which should be changed to "true". Not sure why the `value` property doesn't work. From my understanding this is not a problem right now but let me know if we want to work around that. Thanks!

**Comment by mwt** [comment]
Hi. That makes sense. I think the ideal is for all the fields that are intended to be saved to have a `.saved` class or a `data-saved=true` property so that any input without this property is ignored

Merging the PR actually puts it in a dev environment. I have to sync to the production branch for it to affect the production version.

It's a bit confusing, but:

Dev: https://linkfree.fly.dev

Production: https://linkfree.ckt.im

**Comment by cristianmacedo** [comment]
Ah yes I didn't notice it was the development environment.
That sounds good to me, I will work on a new change to only save to `localStorage` fields with the specified properties (`.saved` class or `data-saved=true`) as soon as possible, and I will be back with a new PR. 
Thanks!

**Pull request merged**